include/mata/nfa/algorithms.hh

```diff
@@ -98,6 +102,22 @@ bool is_universal_naive(const Nfa& aut, const Alphabet& alphabet, Run* cex);
  */
 bool is_universal_antichains(const Nfa& aut, const Alphabet& alphabet, Run* cex);
 
+//the alphabet parameter should be optional? The automaton can have its alphabet too. Run should be optional.
+//actually, what about to use this, with more optional parameters through structures
+//struct univ_params
+//{
+//    Alphabet* alphabet = NULL,
+//    Run* cex = NULL,
+//};
+//   bool is_universal_antichains(const Nfa& aut, univ_params={NULL,NULL});
+//   and call as
+//   bool is_universal_antichains(const Nfa& aut);
+//   bool is_universal_antichains(const Nfa& aut,{.alphabet = bla});
+//   bool is_universal_antichains(const Nfa& aut,{.cex = bli});
+//   bool is_universal_antichains(const Nfa& aut,{.alphabet = bla, .cex = bli});
+
```

Comment:
The failed compilation sounds like som e weird glitch in the matrix, lets pretend that it did not happen?   :)

I somehow think it could be nice with the structs and I guess we would not need so many different ones.

What do the others think?

Reply:
> The failed compilation sounds like som e weird glitch in the matrix, lets pretend that it did not happen? :)

Yeah, Let us. It will be solved when it comes to that. It should not be a significant problem in this case, I hope.

Reply:
Could we employ the proposed key-value structure instead of yet another data type?

Reply:
hm, maybe a good idea? The time needed to fetch the value it should be negligible, right?

Reply:
I am strongly against using a key value store for parameters for operations on the automaton. What would you even do for operations on two automata, each with their own key value store? Using the key value store here would be unreasonable.

Furthermore, if we use the key value store for state-to-state maps (newly `StateRenaming`), Mata would leak memory. Every such operation would require the caller to deallocate the heap-allocated map after the operation. The key-value store should be used only by the user of Mata as a personal storage for their own data, not data generated by the library. Otherwise, the memory leaks will arise, and costly allocations will have to be made repeatedly. Continuous freeing of allocated memory would also be necessary to prevent allocating too much memory at any time during the run of any program using Mata.  

An alternative approach where using key-value store for state renaming would be applicable is to create some sort of global pool of these structures, which will get deallocated when the program ends. Something like that might become a necessity for alphabets (which play a similar role in Mata as data in key-value store memory management-wise). However, the problem of continuous freeing of allocated memory still prevails.  

That would still not warrant the use of the key-value store for optional parameters for operations on automata, however. Using those would be cumbersome at best. We need to pass the parameters truly as parameters to the function, in whatever form that may be, in my opinion.

It would be great if we could design some sort of single key-value store type for optional parameters for all functions. They would specify the type of operation the following parameters are for, and the function would look for the values there in case the parameter of a pointer to this specific key-value store type is not null pointer. Something like this:
```cpp
enum class OperationType {
    Minimize,
    Determinize,
    Trim,
    // ...
}

Nfa& Mata::Nfa::Nfa::minimize(OptionalParameters* optional_parameters = nullptr) {
    OptionalParametersValues* optional_paremeters_values = nullptr;
    if (optional_parameters != nullptr) {
        optional_paremeters_values = optional_parameters[OperationType::Minimize];        
    }
    // ...
    if (optional_parameters_values != nullptr) {
        if (optional_parameters_values->use_reduce) {
             this.reduce();
        }
    }
    return *this;
}
```